### PR TITLE
use the Joi user defined

### PR DESCRIPTION
**Pull request by yujunlong2000**
1: Use the Joi user defined, for example use some extensions for Joi.
2: Use extensions for openapi extensions and add extension for support it.

### CODE REVIEW

**Comment by tlivings** [comment]
Please describe this PR better.

**Comment by yujunlong2000** [comment]
@tlivings I have describe it. can you review it. thanks.

**Comment by tlivings** [comment]
For item 1 - what is the use case for allowing someone to provide an alternative joi library?

**Comment by yujunlong2000** [comment]
@tlivings 
Some times user will extend openapi for some reason, maybe bigint, enum int, but official joi can't express this, so user will extend joi for this, then user should privide that extended joi.

**Comment by tlivings** [comment]
You can already do this through the custom types and refine types features.

File: lib/enjoi.js [hunk 1]
```diff
@@ -286,6 +282,9 @@ const enjoi = function (schema, options = {}) {
 
         Util.isNumber(current.minimum) && (joischema = joischema.min(current.minimum));
         Util.isNumber(current.maximum) && (joischema = joischema.max(current.maximum));
+        Util.isNumber(current.exclusiveMinimum) && (joischema = joischema.greater(current.exclusiveMinimum));
+        Util.isNumber(current.exclusiveMaximum) && (joischema = joischema.less(current.exclusiveMaximum));
+        Util.isNumber(current.multipleOf) && current.multipleOf !== 0 && (joischema = joischema.multiple(current.multipleOf));
```
Comment (tlivings) [inline lib/enjoi.js]: :+1:

File: lib/enjoi.js [hunk 2]
```diff
@@ -286,6 +282,9 @@ const enjoi = function (schema, options = {}) {
 
         Util.isNumber(current.minimum) && (joischema = joischema.min(current.minimum));
         Util.isNumber(current.maximum) && (joischema = joischema.max(current.maximum));
+        Util.isNumber(current.exclusiveMinimum) && (joischema = joischema.greater(current.exclusiveMinimum));
+        Util.isNumber(current.exclusiveMaximum) && (joischema = joischema.less(current.exclusiveMaximum));
```
Comment (tlivings) [inline lib/enjoi.js]: :+1:

File: lib/enjoi.js [hunk 3]
```diff
@@ -286,6 +282,9 @@ const enjoi = function (schema, options = {}) {
 
         Util.isNumber(current.minimum) && (joischema = joischema.min(current.minimum));
         Util.isNumber(current.maximum) && (joischema = joischema.max(current.maximum));
+        Util.isNumber(current.exclusiveMinimum) && (joischema = joischema.greater(current.exclusiveMinimum));
```
Comment (tlivings) [inline lib/enjoi.js]: :+1:

**Comment by yujunlong2000** [comment]
@tlivings I have bisected.

**Pull request closed without merging**